File: example-crs-webservice/crs-userspace/libs/libMSA/python/libmsa/runner.py
from typing import Any, Callable, List, Tuple, Type, TypeVar, Union
import os

from google.protobuf import message as _message

from .kafka.consumer import Consumer
from .kafka.producer import Producer
from .thread.pool import QueuePolicy, ThreadPool

T = TypeVar("T", bound=_message.Message)


class Runner:
    def __init__(
        self,
        input_topic: str,
        input_profobuf_class: Type[T],
        group_id: str,
        output_topic: str | None,
        num_threads: int,
        queue_policy: QueuePolicy,
        func: Callable[
            [_message.Message, int, Any],
            Union[_message.Message, List[_message.Message]],
        ],
        contexts: List[Any],
    ) -> None:
        self.input_topic = input_topic
        self.input_profobuf_class = input_profobuf_class
        self.group_id = group_id
        self.output_topic = output_topic
        self.num_threads = num_threads
        self.queue_policy = queue_policy
        self.func = func
        self.contexts = contexts

        self.kafka_server_addr = os.getenv("KAFKA_SERVER_ADDR")
        assert self.kafka_server_addr

        self.consumer = Consumer(
            self.kafka_server_addr,
            self.input_topic,
            self.group_id,
            self.input_profobuf_class,
        )

        self.producer = (
            Producer(self.kafka_server_addr, self.output_topic)
            if self.output_topic is not None
            else None
        )

        self.thread_pool = ThreadPool(
            self.num_threads, self.queue_policy, self.worker, self.contexts
        )

    def worker(self, data: _message.Message, thread_id: int, context: Any) -> None:
        output_message = self.func(data, thread_id, context)
        if self.producer is not None:
            if isinstance(output_message, list):
                for message in output_message:
                    self.producer.send_message(message)
            else:
                self.producer.send_message(output_message)

    def execute(self) -> None:
        self.thread_pool.execute()
        while True:
            input_message = self.consumer.recv_message()
            if input_message is not None:
                self.thread_pool.enqueue(input_message)

    def execute_thread_pool(self) -> Tuple[Consumer, ThreadPool]:
        self.thread_pool.execute()
        return self.consumer, self.thread_pool


def execute_consumers(consumers: List[Tuple[Consumer, ThreadPool]]) -> None:
    while True:
        for consumer, thread_pool in consumers:
            input_message = consumer.recv_message()
            if input_message is not None:
                thread_pool.enqueue(input_message)

File: example-crs-webservice/crs-userspace/libs/libMSA/rust/libmsa/src/runner.rs
use super::kafka::consumer::Consumer;
use super::kafka::producer::Producer;
use super::thread::pool::QueuePolicy;
use super::thread::pool::ThreadPool;
use prost::Message;
use std::sync::{Arc, Mutex};

pub struct Runner<T, R, C>
where
    T: Message + std::default::Default + Send + Clone + 'static,
    R: Message + std::default::Default + Send + 'static,
    C: Send + Sync + 'static,
{
    input_topic: String,
    group_id: String,
    output_topic: String,
    num_threads: usize,
    queue_policy: QueuePolicy,
    thread_pool: ThreadPool<T, C>,
    consumer: Consumer<T>,
    producer: Arc<Producer<R>>,
}

impl<T, R, C> Runner<T, R, C>
where
    T: Message + std::default::Default + Send + Clone + 'static,
    R: Message + std::default::Default + Send + 'static,
    C: Send + Sync + 'static,
{
    pub fn new<F>(
        input_topic: String,
        group_id: String,
        output_topic: String,
        num_threads: usize,
        queue_policy: QueuePolicy,
        func: F,
        contexts: Option<Vec<Arc<Mutex<C>>>>,
    ) -> Self
    where
        F: Fn(T, usize, Option<Arc<Mutex<C>>>) -> Option<R> + Send + Sync + 'static,
    {
        let kafka_server_addr = std::env::var("KAFKA_SERVER_ADDR").expect("KAFKA_SERVER_ADDR is not set");
        let consumer = Consumer::new(
            kafka_server_addr.clone(),
            input_topic.clone(),
            group_id.clone(),
        );
        let producer = Arc::new(Producer::new(
            kafka_server_addr.clone(),
            output_topic.clone(),
        ));

        let func = Arc::new(func);
        let wrapped_func = {
            let producer = producer.clone();
            let func = func.clone();
            move |input: T, thread_id: usize, context: Option<Arc<Mutex<C>>>| {
                let producer = producer.clone();
                let func = func.clone();
                async move {
                    if let Some(output) = func(input, thread_id, context) {
                        producer.send_message(output).await;
                    }
                }
            }
        };
        let thread_pool = ThreadPool::new(num_threads, queue_policy, wrapped_func, contexts);

        Runner {
            input_topic,
            group_id,
            output_topic,
            num_threads,
            queue_policy,
            thread_pool,
            consumer,
            producer,
        }
    }

    pub async fn execute(&mut self) {
        self.thread_pool.execute();
        loop {
            let input_message = self.consumer.recv_message().await;
            self.thread_pool.enqueue(input_message);
        }
    }
}
